### PR TITLE
add material, light component, edit mesh component

### PR DESCRIPTION
**Pull request by bogdan-voevoda**

### CODE REVIEW

File: editor/rodedit.nim [hunk 1]
```diff
@@ -42,16 +48,40 @@ proc startApplication() =
...
+    let lightMesh = light.component(MeshComponent)
+    lightMesh.loadMeshComponentWithResource("cube.obj")
+    lightMesh.material.setAmbientColor(0.9, 0.9, 0.0)
+    lightMesh.material.isWireframe = true
+    lightMesh.material.removeDiffuseColor()
+    lightMesh.material.removeSpecularColor()
+    lightMesh.material.removeShininess()
+
+    let lightSource = light.component(LightSource)
+    editView.viewport.light = cast[Component](lightSource)
```
Comment (yglukhov) [inline editor/rodedit.nim]: Registration in viewport should occur in LightSource class

File: editor/rodedit.nim [hunk 2]
```diff
@@ -42,16 +48,40 @@ proc startApplication() =
...
+    
+    let mapleTree = editView.viewport.rootNode.newChild("maple_tree")
+    mapleTree.translation = newVector3(-5,-10,-80)
+    let meshTree = mapleTree.component(MeshComponent)
+    meshTree.loadMeshComponentWithResource("tree_maple.obj")
+    meshTree.material.albedoTexture = imageWithResource("tree_maple_color.png")
+    meshTree.material.setDiffuseColor(0.5, 0.5, 0.5)
+    meshTree.material.removeSpecularColor()
+    meshTree.applyLight()
```
Comment (yglukhov) [inline editor/rodedit.nim]: Should be done automagically

File: rod/component/light.nim [hunk 3]
```diff
@@ -0,0 +1,96 @@
+import nimx.types
+import nimx.matrixes
+
+import rod.component
+
+type LightSource* = ref object of Component
+    lightPosition: Vector4
```
Comment (yglukhov) [inline rod/component/light.nim]: Should come from node

File: rod/component/light.nim [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    lightPosInited*: bool
+    lightAmbientInited*: bool
+    lightDiffuseInited*: bool
+    lightSpecularInited*: bool
+    lightConstantInited*: bool
+    lightLinearInited*: bool
+    lightQuadraticInited*: bool
+    lightAttenuationInited*: bool
+
+proc setPosition*(l: LightSource, x, y, z, w: Coord = 0) =
```
Comment (yglukhov) [inline rod/component/light.nim]: Should not be here

File: rod/component/light.nim [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    l.lightLinear = val
+    l.lightLinearInited = true
+proc setQuadratic*(l: LightSource, val: Coord) = 
+    l.lightQuadratic = val
+    l.lightQuadraticInited = true
+proc setAttenuation*(l: LightSource, val: Coord) = 
+    l.lightAttenuation = val
+    l.lightAttenuationInited = true
+
+proc getPosition*(l: LightSource): Vector4 =
```
Comment (yglukhov) [inline rod/component/light.nim]: Should not be here

File: rod/component/light.nim [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    lightSpecularInited*: bool
+    lightConstantInited*: bool
+    lightLinearInited*: bool
+    lightQuadraticInited*: bool
+    lightAttenuationInited*: bool
+
+proc setPosition*(l: LightSource, x, y, z, w: Coord = 0) = 
+    l.lightPosition = newVector4(x, y, z, w)
+    l.lightPosInited = true
+proc setAmbient*(l: LightSource, val: Coord) =
```
Comment (yglukhov) [inline rod/component/light.nim]: consider using the following signature:
```nim
proc `ambient=`(l: LightSource, val: Coord)
```

File: rod/component/light.nim [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+proc setQuadratic*(l: LightSource, val: Coord) = 
+    l.lightQuadratic = val
+    l.lightQuadraticInited = true
+proc setAttenuation*(l: LightSource, val: Coord) = 
+    l.lightAttenuation = val
+    l.lightAttenuationInited = true
+
+proc getPosition*(l: LightSource): Vector4 = 
+    result = l.lightPosition
+proc getAmbient*(l: LightSource): Coord =
```
Comment (yglukhov) [inline rod/component/light.nim]: consider using template instead of proc

File: rod/component/light.nim [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    l.lightAttenuationInited = false
+
+proc newDefaultLightSourceWithPosition*(x, y, z, w: Coord = 0): LightSource = 
+    result.new()
+    result.setDefaultLightSourceWithPosition(x, y, z, w)
+
+method init*(l: LightSource) =
+    l.setDefaultLightSourceWithPosition(0.0, 0.0, 0.0, 0.0)
+    procCall l.Component.init()
```
Comment (yglukhov) [inline rod/component/light.nim]: call super before anything

File: rod/component/light.nim [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+method draw*(l: LightSource) =
```
Comment (yglukhov) [inline rod/component/light.nim]: not needed here.

File: rod/component/light.nim [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    l.setDefaultLightSourceWithPosition(0.0, 0.0, 0.0, 0.0)
+    procCall l.Component.init()
+
+method draw*(l: LightSource) = 
+    l.setPosition(l.node.translation.x, l.node.translation.y, l.node.translation.z, l.lightPosition.w)
+    # l.setAmbient(l.node.scale.x)
+    # l.setDiffuse(l.node.scale.y)
+    # l.setSpecular(l.node.scale.z)
+
+# method componentNodeWasAddedToViewport*(c: Component) {.base.} = discard
```
Comment (yglukhov) [inline rod/component/light.nim]: Register in viewport

File: rod/component/material.nim [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+    result.transform = newTransformInfoWithIdentity()
+
+    result.shader = 0
+    result.shaderNeedUpdate()
+
+proc updateVertexAttributesSetup*(m: Material, vertInfo: VertexDataInfo) =
+    let c = currentContext()
+    let gl = c.gl
+
+    var ofset: int = 0
```
Comment (yglukhov) [inline rod/component/material.nim]: offset

File: rod/component/material.nim [hunk 12]
```diff
@@ -0,0 +1,447 @@
...
+
+    if vertInfo.numOfCoordPerVert != 0: 
+        gl.enableVertexAttribArray(aPosition.GLuint)
+        gl.vertexAttribPointer(aPosition.GLuint, vertInfo.numOfCoordPerVert, gl.FLOAT, false, vertInfo.stride.GLsizei , ofset)
+        ofset += vertInfo.numOfCoordPerVert * sizeof(GLfloat)
+    if vertInfo.numOfCoordPerTexCoord != 0:
+        if m.shader == 0:
+            m.shaderMacroFlags.incl(WITH_V_TEXCOORD)
+        gl.enableVertexAttribArray(aTexCoord.GLuint)
```
Comment (yglukhov) [inline rod/component/material.nim]: why not else?

File: rod/component/material.nim [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+
+proc setupMaterialAttributes(m: Material) = 
+    if not m.color.isNil:
+        let c = currentContext()
+        let gl = c.gl
+
+        if m.color.ambientInited:
+            if m.shader == 0: 
+                m.shaderMacroFlags.incl(WITH_MATERIAL_AMBIENT)
+            gl.uniform4fv(gl.getUniformLocation(m.shader, "uMaterialAmbient"), m.color.ambient)
```
Comment (yglukhov) [inline rod/component/material.nim]: why not else

File: rod/component/material.nim [hunk 14]
```diff
@@ -0,0 +1,447 @@
...
+        m.shaderMacroFlags.incl(WITH_TBN_FROM_NORMALS)
+
+proc updateTransformSetup*(m: Material, translation: Vector3, rotation: Quaternion, scale: Vector3) =
+    m.transform.fromScaleRotationTranslation(scale, rotation, translation)
+
+proc createShader(m: Material) = 
+    let c = currentContext()
+    let gl = c.gl
+
+    #TODO delete shader prog from gl if exist
```
Comment (yglukhov) [inline rod/component/material.nim]: delete it ;)

File: rod/component/material.nim [hunk 15]
```diff
@@ -0,0 +1,447 @@
...
+    # useNormals: bool
+    # useNormalMap: bool
+    # useLight: bool
+    # TODO
+
+    blendEnable*: bool
+    depthEnable*: bool
+    isWireframe*: bool
+    
+    vertexShader*: string
```
Comment (yglukhov) [inline rod/component/material.nim]: private

File: rod/component/material.nim [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+
+    #TODO delete shader prog from gl if exist 
+    m.shader = 0
+
+    var commonShaderDefines = ""
+    for macros in m.shaderMacroFlags:
+        commonShaderDefines &= """#define """ & $macros & "\n"
+
+    if m.vertexShader.len == 0:
+        m.vertexShader = ""
```
Comment (yglukhov) [inline rod/component/material.nim]: m.vertexShader = commonShaderDefines & materialVertexShaderDefault

File: rod/component/mesh_component.nim [hunk 17]
```diff
@@ -1,10 +1,193 @@
...
+    numberOfIndices*: GLsizei
+    loadFunc: proc()
+    vertInfo*: VertexDataInfo
+    material*: Material
+
+method init*(m: MeshComponent) =
+    m.material = newDefaultMaterial()
+    procCall m.Component.init()
+
+proc applyLight*(m: MeshComponent) =
```
Comment (yglukhov) [inline rod/component/mesh_component.nim]: remove this

File: rod/component/mesh_component.nim [hunk 18]
```diff
@@ -1,10 +1,193 @@
...
+    material*: Material
+
+method init*(m: MeshComponent) =
+    m.material = newDefaultMaterial()
+    procCall m.Component.init()
+
+proc applyLight*(m: MeshComponent) = 
+    m.material.light = cast[LightSource](m.node.mViewport.light)
+
+proc setInstansedVBOAttributes*(m: MeshComponent, indBuffer, vertBuffer: GLuint, numOfIndices: GLsizei, vInfo: VertexDataInfo) =
```
Comment (yglukhov) [inline rod/component/mesh_component.nim]: instanCed

File: rod/component/mesh_component.nim [hunk 19]
```diff
@@ -1,10 +1,193 @@
...
+        if currentContext().isNil:
+            m.loadFunc = loadFunc
+        else:
+            loadFunc()
+
+proc loadMeshComponentWithResource*(m: MeshComponent, resourceName: string) =
+    m.loadFunc = proc() =
+        m.loadMeshComponent(resourceName)
+
+proc newMeshComponentWithResource*(resourceName: string): MeshComponent =
```
Comment (yglukhov) [inline rod/component/mesh_component.nim]: Should be removed.

File: rod/component/mesh_component.nim [hunk 20]
```diff
@@ -1,10 +1,193 @@
...
+    m.vertexBuffer = gl.createBuffer()
+    gl.bindBuffer(gl.ARRAY_BUFFER, m.vertexBuffer)
+    gl.bufferData(gl.ARRAY_BUFFER, vertexData, gl.STATIC_DRAW)
+    m.numberOfIndices = indexData.len.GLsizei
+
+proc meshComponentWithQuad*(m: MeshComponent, v1, v2, v3, v4: Vector3, t1, t2, t3, t4: Point) =
+    m.loadFunc = proc() =
+        m.loadMeshQuad(v1, v2, v3, v4, t1, t2, t3, t4)
+
+proc newMeshComponentWithQuad*(v1, v2, v3, v4: Vector3, t1, t2, t3, t4: Point): MeshComponent =
```
Comment (yglukhov) [inline rod/component/mesh_component.nim]: should be removed

File: rod/component/mesh_component.nim [hunk 21]
```diff
@@ -1,10 +1,193 @@
...
+    resourceName: string
+    indexBuffer*: GLuint
+    vertexBuffer*: GLuint
+    numberOfIndices*: GLsizei
+    loadFunc: proc()
+    vertInfo*: VertexDataInfo
+    material*: Material
+
+method init*(m: MeshComponent) =
+    m.material = newDefaultMaterial()
```
Comment (yglukhov) [inline rod/component/mesh_component.nim]: defer material creation to mesh loading

File: rod/mesh.nim [hunk 22]
```diff
@@ -16,15 +15,13 @@ import streams
 type Attr = enum
     aPosition
     aTexCoord
+    aNormal
```
Comment (yglukhov) [inline rod/mesh.nim]: Should be rolled back and deprecated. Please fix the bug in nimasset to make it compatible with this Mesh impl

File: rod/node.nim [hunk 23]
```diff
@@ -49,6 +51,10 @@ type
         mActiveFrameBuffer*, mBackupFrameBuffer*: SelfContainedImage
         mScreenFrameBuffer*: GLuint
         tempFramebuffers*: seq[SelfContainedImage]
+        # passID
+        # renderPath
+        # observ
+        light*: Component
```
Comment (yglukhov) [inline rod/node.nim]: Add a coment about nims annoying cyclic dependencies

**Comment by yglukhov** [comment]
Please test in JS!

File: rod/node.nim [hunk 24]
```diff
@@ -15,6 +15,8 @@ import nimx.view
 import quaternion
 import property_visitor
 
+# import rod.component.light
+
 type
```
Comment (yglukhov) [inline rod/node.nim]: Move this type section out to rod_types.nim? And add LightSource to it.

File: rod/vertex_data_info.nim [hunk 25]
```diff
@@ -0,0 +1,30 @@
+import nimx.portable_gl
+
+type VertexDataInfo* = ref object of RootObj
```
Comment (yglukhov) [inline rod/vertex_data_info.nim]: Should not be ref

File: rod/component/material.nim (hunk 12 above)
Comment (bogdan-voevoda) [inline rod/component/material.nim]: this procedure calls from mesh_component.nim once per draw call, so vertex attributes must be initialised before shader prog construction

**Comment by yglukhov** [comment]
Thank you!

**Pull request merged**